---
title: "Relative Proportion of Intermediate Methylation (RPIM)"
author: "Nathan Sheffield & VP Nagraj"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Relative Proportion of Intermediate Methylation (RPIM)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = F}
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE)
```


## Overview

RPIM is a score that measures the epigenetic heterogeneity in a bisulfite sequencing sample. Under the assumption that a homogeneous sample will have mostly CpGs with either 100 it follows that the proportion of sites that differ from these two extremes can be used as a measure of sample heterogeneity.

```{r}

```

## Data Preparation

RPIM requires DNA methylation data after methylation calling. For a given genomic coordinate (the location of the C in a CpG), RPIM needs two values: the number of methylated reads, and the total number of reads. This data should be represented as a `data.table` for each sample, which we call a BSDT (Bisulfite data.table). The BSDT will have these columns: `chr`, `start` (the coordinate of the C of the CpG), `methylCount` (number of methylated reads), `coverage` (total number of reads covering this site), `methylProp` (`methylCount/coverage`), and (optionally if you have one sample but required if you have multiple samples) `sampleName` (sample identifier). Since some existing R packages for DNA methylation use different formats, we include a format conversion function that can be used to convert `SummarizedExperiment`-based objects like you would obtain from the `bsseq`, `methylPipe`, and `BiSeq` packages to the necessary format for MIRA (`SummarizedExperimentToDataTable` function). Here is an example of a `data.table` in the right format for input to MIRA:

```{r}
data("exampleBSDT", package="MIRA")
head(exampleBSDT)
```

## Calculating PIM

## Calculating RPIM

```{r, eval = F}
# devtools::install_github("databio/MIRA")
# devtools::install_github("databio/simpleCache")

library(RPIM)

dat = MIRA::BSreadBiSeq("data/RRBS_cpgMethylation_EWS_L10.bed")

# have to setSharedCacheDir()
simpleCache::setSharedCacheDir("cache")

imres = calculatePIM(dat)

###############################

# relative proportion of sites

dat2 = MIRA::BSreadBiSeq("data/RRBS_cpgMethylation_EWS_T133.bed")
dat3 = MIRA::BSreadBiSeq("data/RRBS_cpgMethylation_EWS_T111.bed")
dat4 = MIRA::BSreadBiSeq("data/RRBS_cpgMethylation_EWS_T120.bed")

alldat = rbind(dat,dat2, dat3, dat4)
allsplitdat = split(alldat, alldat$sampleName)

getRPIM(allsplitdat)

# try this with bsseq object
# source("https://bioconductor.org/biocLite.R")
# biocLite("bsseqData")

library(bsseqData)
data(BS.cancer.ex)

bsseqdat = MIRA::bsseqToDataTable(BS.cancer.ex)

getRPIM(bsseqdat, cache = TRUE, cacheDir = "baz2")

getRPIM(bsseqdat, cache = FALSE)
```

